File: AikumaCloudStorage/src/org/lp20/aikuma/storage/google/GoogleDriveIndex2.scala
package org.lp20.aikuma.storage.google

import java.util.{Map => JMap}
import org.lp20.aikuma.net.Http
import org.lp20.aikuma.storage.Utils

class GoogleDriveIndex2(rootTitle: String, tm: TokenManager, baseUrl: String, idToken: String)
extends GoogleDriveIndex(rootTitle, tm) {

  override def index(identifier: String, metadata: JMap[String,String]) =
    call("POST", identifier, metadata)

  override def update(identifier: String, metadata: JMap[String,String]) =
    call("PUT", identifier, metadata)

  private def call(method: String, identifier: String, metadata: JMap[String,String]): Boolean = {
    if (identifier == null || metadata == null)
      return false

    Http(s"$baseUrl/$identifier")
      .method(method)
      .header("X-Aikuma-Auth-Token", idToken)
      .body(Utils.map2query(metadata))
      .code match {
        case 202 => true
        case _ => false
      }
  }
}

File: AikumaCloudStorage/src/org/lp20/aikuma/storage/google/GoogleDriveIndex2.java
package org.lp20.aikuma.storage.google;

import java.io.IOException;
import java.net.MalformedURLException;
import java.net.ProtocolException;
import java.util.Map;

import org.lp20.aikuma.net.Http;
import org.lp20.aikuma.storage.Utils;
import org.lp20.aikuma.storage.DataStore;

public class GoogleDriveIndex2 extends GoogleDriveIndex {

	private String baseUrl;
	private String idToken;
	
	public GoogleDriveIndex2(String rootTitle, TokenManager tm, String baseUrl, String idToken) 
			throws DataStore.StorageException {
		super(rootTitle, tm);
		this.baseUrl = baseUrl;
		this.idToken = idToken;
	}
	
	public boolean index(String identifier, Map<String, String> metadata) {
		return call("POST", identifier, metadata);
	}
	
	public boolean update(String identifier, Map<String, String> metadata) {
		return call("PUT", identifier, metadata);
	}
	
	private boolean call(String method, String identifier, Map<String, String> metadata) {
		if(identifier == null || metadata == null)
			return false;
		
		String url = String.format("%s/%s", baseUrl, identifier);
		Http http = null;
		try {
			http = Http.apply(url)
						  .method(method)
						  .header("X-Aikuma-Auth-Token", idToken)
						  .header("Content-Type", "application/x-www-form-urlencoded")
						  .body(Utils.map2query(metadata));
		} catch (ProtocolException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (MalformedURLException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
		if(http != null && http.code() == 202)
			return true;
		else
			return false;
	}
}